corehq/project_limits/tests/test_rate_definition.py

```diff
@@ -0,0 +1,169 @@
+
+
+def test_rate_definition_map():
+    def check(rate_def, fn):
+        testil.eq(
+            rate_def.map(fn),
+            RateDefinition(
+                per_second=fn(rate_def.per_second),
+                per_minute=fn(rate_def.per_minute),
+                per_hour=fn(rate_def.per_hour),
+                per_day=fn(rate_def.per_day),
+                per_week=fn(rate_def.per_week),
+            ),
+        )
+
+    def preserve_none(fn):
+        return lambda x: None if x is None else fn(x)
+
+    for _ in range(15):
+        rate_def = _get_random_rate_def()
+        fn, = random.choices([preserve_none(int),
+                              preserve_none(lambda x: x * x),
+                              preserve_none(math.exp)])
+        yield check, rate_def, fn
+
+
+def test_rate_definition_map_with_other():
+    def check(rate_def, fn, other):
+        testil.eq(
+            rate_def.map(fn, other),
+            RateDefinition(
+                per_second=fn(rate_def.per_second, other.per_second),
+                per_minute=fn(rate_def.per_minute, other.per_minute),
+                per_hour=fn(rate_def.per_hour, other.per_hour),
+                per_day=fn(rate_def.per_day, other.per_day),
+                per_week=fn(rate_def.per_week, other.per_week),
+            ),
+        )
+
+    for _ in range(15):
+        rate_def = _get_random_rate_def()
+        fn, = random.choices([lambda x, y: max(filter(None, [x, y, -1])),
+                              lambda x, y: None if None in (x, y) else (x * y)])
+        other = _get_random_rate_def()
+        yield check, rate_def, fn, other
+
+
+def _get_random_rate_def():
+    numbers = random.choices([None] + [random.expovariate(1 / 64) for _ in range(6)], k=5)
+    return RateDefinition(
+        per_second=numbers[0],
+        per_minute=numbers[1],
+        per_hour=numbers[2],
+        per_day=numbers[3],
+        per_week=numbers[4],
+    )
+
+
+def test_rate_definition_against_fixed_user_table():
+    tab_delimited_table = """
+    # of users  second  minute  hour    day       week
+    1           1       10      30      50        115
+    10          1       10      30      230       1,150
+    100         1       10      300     2,300     11,500
+    1000        5       70      3,000   23,000    115,000
+    10000       50      700     30,000  230,000   1,150,000
+    100000      500     7,000   300,000 2,300,000 11,500,000
+    """
+
+    rows = tab_delimited_table.strip().splitlines()[1:]
+    table = [[int(cell.replace(',', '')) for cell in row.split()] for row in rows]
+    per_user_rate = RateDefinition(
+        per_week=115,
+        per_day=23,
+        per_hour=3,
+        per_minute=0.07,
+        per_second=0.005,
+    )
+    floor_rate = RateDefinition(
+        per_day=50,
+        per_hour=30,
+        per_minute=10,
+        per_second=1,
+    )
+    for n_users, per_second, per_minute, per_hour, per_day, per_week in table:
```

Comment:
Could also use a generator test to make each row into a separate test.

Reply:
Good catch. Tacking on here: https://github.com/dimagi/commcare-hq/pull/25395/commits/3b1666f1f92d762b688501fea4d3bd348c3c644c